### PR TITLE
KAFKA-6718 / Update SubscriptionInfoData with clientTags

### PR DESCRIPTION
**Pull request by lkokhreidze**
This PR is Part of [KIP-708](https://cwiki.apache.org/confluence/display/KAFKA/KIP-708%3A+Rack+awareness+for+Kafka+Streams), which adds `ClientTags` to `SubscriptionInfoData`. 

Splitting PRs into three smaller PRs to make the review process easier to follow. Overall plan is the following:

⏭️  Rack aware standby task assignment logic. https://github.com/apache/kafka/pull/10851
👉  Protocol change, add `clientTags` to `SubscriptionInfoData`
⏭️  Add required configurations to `StreamsConfig` (public API change, at this point we should have full functionality) https://github.com/apache/kafka/pull/11837

This PR adds `clientTags` to `SubscriptionInfoData` and implements second point of above mentioned plan.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by lkokhreidze** [comment]
Call for review @cadonna @vvcephei @ableegoldman

**Comment by lkokhreidze** [comment]
Hi @cadonna @vvcephei @ableegoldman, I know it's busy times with 3.0 release on the radar, but would appreciate review on this PR when you got time. Thanks!

File: streams/src/main/resources/common/message/SubscriptionInfoData.json
Comment (cadonna) [inline streams/src/main/resources/common/message/SubscriptionInfoData.json]: I think this should be 11+. While technically it probably does not make any difference, it better documents when the struct was introduced.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/SubscriptionInfo.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/SubscriptionInfo.java]: nit:
```suggestion
    private List<ClientTag> buildClientTagsFromMap(final Map<String, String> clientTags) {
        return clientTags.entrySet().stream()
            .map(clientTagEntry -> {
                final ClientTag clientTag = new ClientTag();
                clientTag.setKey(clientTagEntry.getKey().getBytes(StandardCharsets.UTF_8));
                clientTag.setValue(clientTagEntry.getValue().getBytes(StandardCharsets.UTF_8));
                return clientTag;
            })
            .collect(Collectors.toList());
    }
```

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/SubscriptionInfo.java
Comment (cadonna) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/SubscriptionInfo.java]: nit:
```suggestion
        return data.clientTags().stream()
            .collect(
                Collectors.toMap(
                    clientTag -> new String(clientTag.key(), StandardCharsets.UTF_8),
                    clientTag -> new String(clientTag.value(), StandardCharsets.UTF_8)
                )
            );
```

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/SubscriptionInfoTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/SubscriptionInfoTest.java]: Could you use a static final variable named `IGNORED_CLIENT_TAGS` to better document the code as was done for some other fields?
Here and below.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/SubscriptionInfoTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/SubscriptionInfoTest.java:476]: Could you add a test to verify what happens when an empty client tags map is passed to a version 11 subscription info?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/SubscriptionInfoTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/SubscriptionInfoTest.java]: Could you use a map with more than just one entry? If you use the same map in multiple tests, you should put it into a class field. The same applies to the tests below.

File: tests/kafkatest/tests/streams/streams_upgrade_test.py
Comment (cadonna) [inline tests/kafkatest/tests/streams/streams_upgrade_test.py]: Awesome that you thought about this!

File: streams/src/test/java/org/apache/kafka/streams/tests/StreamsUpgradeTest.java
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/tests/StreamsUpgradeTest.java]: Could you put the map into variable with a meaningful name?
Would make sense to use a non-empty map here?

**Review by cadonna — COMMENTED** [review]
@lkokhreidze Thanks for the PR!

I reviewed this PR and we bring this to an approved state. However, I was thinking that maybe it would be better to merge the standby task assignor first, since we do not know if we will merge this PR before 3.0 is cut. If we do not manage to merge it then we bumped the version of the subscription for nothing. I think bumping the subscription version has some risks and increases the number of rebalances in a rolling upgrade due to the whole version probing mechanism that relies on that version.
 
Here is my feedback!

**Comment by lkokhreidze** [comment]
Hi @cadonna,
Thanks for the feedback. I think that's a good call. I didn't know that consequence of bumping the version. I will open PR around assignor changes shortly, and re-brand this PR as "Part 2"

File: streams/src/test/java/org/apache/kafka/streams/tests/StreamsUpgradeTest.java
Comment (lkokhreidze) [reply streams/src/test/java/org/apache/kafka/streams/tests/StreamsUpgradeTest.java]: Makes sense. Added non-empty map as a static final field.

**Comment by lkokhreidze** [comment]
Hi @cadonna I've addressed all of your comments.

**Comment by lkokhreidze** [comment]
Hi @cadonna 
Giving it a bit more thought around the order of PRs, logically it makes more sense to have this PR first, as TaskAssignor gets data from the subscription info. 
Coming back to your point about bumping the version increasing the number of rebalances in a rolling upgrade scenario - considering that protocol version for the 3.0 was already increased via https://github.com/apache/kafka/pull/10609, do you think it's still a problem?
I would prefer to finalise this PR first and avoid more context switching, but of course if it's really needed I can switch to task assignor implementation.

Thanks again for the feedback!

**Comment by lkokhreidze** [comment]
Hi @cadonna
FYI, I've opened a PR around rack aware standby task assignment https://github.com/apache/kafka/pull/10851

**Comment by lkokhreidze** [comment]
Hi @cadonna @showuon 
Sorry for the ping, but was wondering if it's possible to have a look at this PR this week?
After this is merged, I'll have all the pieces to glue everything together in the last PR of this KIP https://github.com/apache/kafka/pull/11837

Thanks.

**Comment by cadonna** [comment]
@lkokhreidze Thank you for the ping! I will plan to look at your PR this week. Let's try to get the whole feature merged before feature freeze for AK 3.2.0 which is March 16th!

**Review by cadonna — APPROVED** [review]
Thanks for the PR @lkokhreidze !

LGTM!

I started a system test run to double-check the behavior.

**Comment by showuon** [comment]
I'll start to take a look this week. Sorry for late response.

**Comment by cadonna** [comment]
@lkokhreidze Could you please rebase this PR on latest trunk? Some system tests fail probably due to the absence of a fix on the PR branch.

**Comment by lkokhreidze** [comment]
@cadonna done!

**Comment by cadonna** [comment]
@lkokhreidze The system tests passed:

http://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/2022-03-10--001.system-test-kafka-branch-builder--1646904256--lkokhreidze--KAFKA-6718-part1-subscription-info-changes--3e73e5b962/report.html

@showuon Could you merge this PR after your review? (Of course, only if the PR looks good to you)

File: tests/kafkatest/tests/streams/streams_upgrade_test.py
Comment (showuon) [inline tests/kafkatest/tests/streams/streams_upgrade_test.py:480]: Thanks for remembering the system tests!

**Review by showuon — APPROVED** [review]
LGTM! Thanks for splitting the KIP into smaller PRs.

**Comment by showuon** [comment]
Failed tests are unrelated:
```
    Build / JDK 11 and Scala 2.13 / org.apache.kafka.connect.mirror.integration.MirrorConnectorsIntegrationSSLTest.testOffsetSyncsTopicsOnTarget()
    Build / JDK 11 and Scala 2.13 / org.apache.kafka.connect.integration.ConnectorClientPolicyIntegrationTest.testCreateWithNoAllowedOverridesForNonePolicy
    Build / JDK 11 and Scala 2.13 / org.apache.kafka.connect.integration.RebalanceSourceConnectorsIntegrationTest.testDeleteConnector
    Build / JDK 11 and Scala 2.13 / org.apache.kafka.connect.integration.SinkConnectorsIntegrationTest.testEagerConsumerPartitionAssignment
    Build / JDK 11 and Scala 2.13 / kafka.api.ConsumerBounceTest.testClose()
    Build / JDK 11 and Scala 2.13 / kafka.api.ConsumerBounceTest.testClose()
    Build / JDK 11 and Scala 2.13 / kafka.server.KRaftClusterTest.testCreateClusterAndPerformReassignment()
    Build / JDK 17 and Scala 2.13 / org.apache.kafka.connect.integration.SourceConnectorsIntegrationTest.testSwitchingToTopicCreationEnabled
    Build / JDK 17 and Scala 2.13 / kafka.api.ConsumerBounceTest.testClose()
    Build / JDK 8 and Scala 2.12 / kafka.api.ConsumerBounceTest.testClose()
    Build / JDK 8 and Scala 2.12 / kafka.api.ConsumerBounceTest.testClose()
    Build / JDK 11 and Scala 2.13 / kafka.server.DynamicBrokerReconfigurationTest.testThreadPoolResize()
    Build / JDK 11 and Scala 2.13 / kafka.server.DynamicBrokerReconfigurationTest.testThreadPoolResize()
    Build / JDK 17 and Scala 2.13 / kafka.server.DynamicBrokerReconfigurationTest.testThreadPoolResize()
    Build / JDK 17 and Scala 2.13 / kafka.server.DynamicBrokerReconfigurationTest.testThreadPoolResize()
```

**Pull request merged**
